FirebaseInstallations/Source/Library/FIRInstallations.m

```diff
@@ -143,6 +143,40 @@ + (void)validateAppOptions:(FIROptions *)appOptions appName:(NSString *)appName
             kFIRLoggerInstallations, kFIRInstallationsMessageCodeInvalidFirebaseAppOptions,
             [missingFields componentsJoinedByString:@", "]];
   }
+
+  [self validateAPIKey:appOptions.APIKey];
+}
+
++ (void)validateAPIKey:(nullable NSString *)APIKey {
+  NSMutableArray<NSString *> *validationIssues = [[NSMutableArray alloc] init];
+
+  if (APIKey.length != kExpectedAPIKeyLength) {
+    [validationIssues addObject:[NSString stringWithFormat:@"API Key length must be %lu characters",
+                                                           (unsigned long)kExpectedAPIKeyLength]];
+  }
+
+  if (![[APIKey substringToIndex:1] isEqualToString:@"A"]) {
+    [validationIssues addObject:@"API Key must start with `A`"];
+  }
+
+  NSCharacterSet *characters = [NSCharacterSet characterSetWithCharactersInString:APIKey];
+  if (![[NSCharacterSet alphanumericCharacterSet] isSupersetOfSet:characters]) {
+    [validationIssues addObject:@"API Key must contain only alphanumeric characters"];
+  }
```

Comment:
Out of curiosity:
Why can't you use RegEx here?
You said something about RegEx being slow on iOS?
RegEx seem to be easier to read and thus prevent implementation errors.

Reply:
This validation happens on the main thread at a critical part of iOS app startup, so we have optimize it. We had to get rid of RegEx for appID validation at some point because of the same reason https://github.com/firebase/firebase-ios-sdk/pull/2529. If it was done at another place on a background thread, we would probably use RegEx (though I'm not a big fan of them).

Reply:
This sounds weird to me.
I don't know of the regular expression implementation on iOS, but usually a well written regular expression is more efficient than the same checks in imperative code, because regular expressions can be precompiled and optimized before the program is actually run.
Maybe the regEx engine is not good on iOS? Well, who knows.

Reply:
I wish iOS provided API to precompile regEx, but I'm not aware of such. Instantiating NSRegularExpression object with a pattern (the [method](https://developer.apple.com/documentation/foundation/nsregularexpression/1410900-init)) was the expensive part. Evaluation was not an issue in #2529.

Reply:
But you convinced me to measure the performance diff between the current and the regEx implementation for this case. Will do it and update if regEx performs well here.